extensions/emmet/src/abbreviationActions.ts

```diff
@@ -59,55 +47,95 @@ function doWrapping(_: boolean, args: any) {
+					(startNode.close && startOffset > startNode.close.start && startOffset < startNode.close.end)
+				)
+			) {
+				start = document.positionAt(startNode.start);
+				const nodeEndPosition = document.positionAt(startNode.end);
+				end = nodeEndPosition.isAfter(end) ? nodeEndPosition : end;
+			}
+
+			const endOffset = document.offsetAt(end);
+			const endNode = getFlatNode(rootNode, endOffset, true) as (HtmlNode | undefined);
+			if (
+				endNode &&
+				(
+					(endNode.open && endOffset > endNode.open.start && endOffset < endNode.open.end) ||
+					(endNode.close && endOffset > endNode.close.start && endOffset < endNode.close.end)
+				)
+			) {
+				const nodeStartPosition = document.positionAt(endNode.start);
+				start = nodeStartPosition.isBefore(start) ? nodeStartPosition : start;
+				const nodeEndPosition = document.positionAt(endNode.end);
+				end = nodeEndPosition.isAfter(end) ? nodeEndPosition : end;
+			}
+
+			rangeToReplace = new vscode.Range(start, end);
+		}
+		// in case of multi-line, exclude last empty line from rangeToReplace
 		if (!rangeToReplace.isSingleLine && rangeToReplace.end.character === 0) {
-			// in case of multi-line, exclude last empty line from rangeToReplace
 			const previousLine = rangeToReplace.end.line - 1;
-			const lastChar = document.lineAt(previousLine).text.length;
-			rangeToReplace = new vscode.Range(rangeToReplace.start, new vscode.Position(previousLine, lastChar));
-		} else if (rangeToReplace.isEmpty) {
-			const { active } = selection;
-			const activeOffset = document.offsetAt(active);
-			const currentNode = getFlatNode(rootNode, activeOffset, true);
-			if (currentNode) {
-				const currentNodeStart = document.positionAt(currentNode.start);
-				const currentNodeEnd = document.positionAt(currentNode.end);
-				if (currentNodeStart.line === active.line || currentNodeEnd.line === active.line) {
-					// wrap around entire node
-					rangeToReplace = new vscode.Range(currentNodeStart, currentNodeEnd);
-				}
-				else {
-					// wrap line that cursor is on
-					rangeToReplace = new vscode.Range(rangeToReplace.start.line, 0, rangeToReplace.start.line, document.lineAt(rangeToReplace.start.line).text.length);
-				}
+			rangeToReplace = new vscode.Range(rangeToReplace.start, document.lineAt(previousLine).range.end);
+		}
+		// wrap line the cursor is on
+		if (rangeToReplace.isEmpty) {
+			rangeToReplace = document.lineAt(rangeToReplace.start).range;
+		}
+
+		// ignore whitespace on the first line
+		const firstLineOfRange = document.lineAt(rangeToReplace.start);
+		if (!firstLineOfRange.isEmptyOrWhitespace && firstLineOfRange.firstNonWhitespaceCharacterIndex > rangeToReplace.start.character) {
+			rangeToReplace = rangeToReplace.with(new vscode.Position(rangeToReplace.start.line, firstLineOfRange.firstNonWhitespaceCharacterIndex));
+		}
+
+		return rangeToReplace;
+	}).reduce((mergedRanges, range, idx) => {
+		// Merge overlapping ranges
+		if (idx > 0) {
```

Comment:
What's this condition for?

Reply:
The first element (`idx === 0`) should always be added to `mergedRanges` as it starts as an empty array